src/alchemlyb/estimators/bar_.py

```diff
@@ -0,0 +1,135 @@
+        Lambda states for which free energy differences were obtained.
+
+    """
+
+    def __init__(self, maximum_iterations=10000, relative_tolerance=1.0e-7, method='false-position', verbose=False):
+
+        self.maximum_iterations = maximum_iterations
+        self.relative_tolerance = relative_tolerance
+        self.method = method
+        self.verbose = verbose
+
+        # handle for pymbar.BAR object
+        self._bar = None
+
+    def fit(self, u_nk):
+        """
+        Compute overlap matrix of reduced potentials using
+        Bennett acceptance ratio.
+
+        Parameters
+        ----------
+        u_nk : DataFrame 
+            u_nk[n,k] is the reduced potential energy of uncorrelated
+            configuration n evaluated at state k.
+
+        """
+        # sort by state so that rows from same state are in contiguous blocks
+        u_nk = u_nk.sort_index(level=u_nk.index.names[1:])
+
+        # get a list of the lambda states
+        self.states_ = u_nk.columns.values.tolist()
+
+        # group u_nk by lambda states
+        groups = u_nk.groupby(level=u_nk.index.names[1:])
+        N_k = [(len(groups.get_group(i)) if i in groups.groups else 0) for i in u_nk.columns]
+
+        # Now get free energy differences and their uncertainties for each step
+        deltas = np.array([])
+        d_deltas = np.array([])
+        for k in range(len(N_k) - 1):
+            # get us from lambda step k
+            uk = groups.get_group(self.states_[k])
+            # get w_F
+            w_f = uk.iloc[:, k+1] - uk.iloc[:, k]
+
+            # get us from lambda step k+1
+            uk1 = groups.get_group(self.states_[k+1])
+            # get w_R
+            w_r = uk1.iloc[:, k] - uk1.iloc[:, k+1]
+
+            # now determine df and ddf using pymbar.BAR
+            df, ddf = BAR_(w_f, w_r,
+                             method=self.method,
+                             maximum_iterations=self.maximum_iterations,
+                             relative_tolerance=self.relative_tolerance,
+                             verbose=self.verbose)
+
+            deltas = np.append(deltas, df)
+            d_deltas = np.append(d_deltas, ddf**2)
+
+        # build matrix of deltas between each state
+        adelta = np.zeros((len(deltas) + 1, len(deltas) + 1))
+        ad_delta = np.zeros_like(adelta)
+
+        for j in range(len(deltas)):
+            out = []
+            dout = []
+            for i in range(len(deltas) - j):
+                out.append(deltas[i:i + j + 1].sum())
+
+                # See https://github.com/alchemistry/alchemlyb/pull/60#issuecomment-430720742
+                # Error estimate generated by BAR ARE correlated
+
+                # Use the BAR uncertainties between two neighbour states
+                if j == 0:
+                    dout.append(d_deltas[i:i + j + 1].sum())
+                # Other uncertainties are unknown at this point
+                else:
+                    dout.append(float('NaN'))
```

Comment:
Or use `np.NaN` directly, given that we already use numpy. Not sure what's more Pythonic although [float("nan") seems to be the way to go without numpy](https://stackoverflow.com/a/19374296).

Bogus performance "argument" in favor of numpy:
```python
In [17]: %time float("NaN")
CPU times: user 4 µs, sys: 0 ns, total: 4 µs
Wall time: 9.06 µs
Out[17]: nan

In [18]: %time np.NaN
CPU times: user 4 µs, sys: 0 ns, total: 4 µs
Wall time: 7.87 µs
Out[18]: nan
```